File: app/src/main/java/ru/neverlands/abclient/utils/FileLogger.java
package ru.neverlands.abclient.utils;

import android.util.Log;

import java.io.File;
import java.io.FileOutputStream;
import java.io.OutputStreamWriter;
import java.nio.charset.Charset;
import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.Locale;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

public final class FileLogger {
    private static final String TAG = "FileLogger";
    private static final Charset UTF8 = Charset.forName("UTF-8");
    private static final SimpleDateFormat DAY_FORMAT = new SimpleDateFormat("yyyyMMdd", Locale.US);
    private static final SimpleDateFormat TS_FORMAT = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss.SSS", Locale.US);
    private static final SimpleDateFormat PROXY_SEGMENT_FORMAT = new SimpleDateFormat("yyyyMMdd_HH", Locale.US);
    private static final Object FILE_LOCK = new Object();
    private static final long MAX_FILE_BYTES = 8L * 1024L * 1024L;
    private static final int MAX_ROTATIONS = 2;
    private static final ExecutorService IO = Executors.newSingleThreadExecutor(r -> {
        Thread thread = new Thread(r, "ab-file-logger");
        thread.setDaemon(true);
        return thread;
    });

    private FileLogger() {
    }

    public static void trace(String chain, String message) {
        write("TRACE", chain, message, null);
    }

    public static void warn(String chain, String message) {
        write("WARN", chain, message, null);
    }

    public static void error(String chain, String message, Throwable error) {
        write("ERROR", chain, message, error);
    }

    /**
     * Логирует proxy-pool трафик в фиксированный файл `files/Logs/pool/*_proxy.txt`.
     * Используется для детального post/get трассинга локального proxy-контура.
     */
    public static void proxyPool(String message) {
        writeToProxySegment("TRACE", message, null);
    }

    public static void proxyPoolError(String message, Throwable error) {
        writeToProxySegment("ERROR", message, error);
    }

    private static void write(String level, String chain, String message, Throwable error) {
        if (chain == null || chain.trim().isEmpty() || message == null || message.trim().isEmpty()) {
            return;
        }
        final String safeLevel = level == null ? "TRACE" : level.trim().toUpperCase(Locale.ROOT);
        final String safeChain = sanitizeChain(chain);
        final String safeMessage = sanitizeMessage(message);
        IO.execute(() -> appendLine(safeLevel, safeChain, safeMessage, error));
    }

    private static void writeToProxySegment(String level, String message, Throwable error) {
        if (message == null || message.trim().isEmpty()) {
            return;
        }
        final String safeLevel = level == null ? "TRACE" : level.trim().toUpperCase(Locale.ROOT);
        final String safeMessage = sanitizeMessage(message);
        IO.execute(() -> appendLineToProxySegment(safeLevel, safeMessage, error));
    }

    private static void appendLine(String level, String chain, String message, Throwable error) {
        try {
            File logsRoot = resolveLogsRoot();
            if (logsRoot == null) {
                return;
            }
            File criticalDir = new File(logsRoot, "Critical");
            if (!criticalDir.exists() && !criticalDir.mkdirs()) {
                Log.e(TAG, "Failed to create critical log directory: " + criticalDir.getAbsolutePath());
                return;
            }

            String day = DAY_FORMAT.format(new Date());
            File targetFile = new File(criticalDir, day + "_" + chain + ".log");
            synchronized (FILE_LOCK) {
                rotateIfNeeded(targetFile);
                try (FileOutputStream stream = new FileOutputStream(targetFile, true);
                     OutputStreamWriter writer = new OutputStreamWriter(stream, UTF8)) {
                    writer.write(TS_FORMAT.format(new Date()));
                    writer.write(" [");
                    writer.write(level);
                    writer.write("] ");
                    writer.write(message);
                    writer.write('\n');
                    if (error != null) {
                        writer.write(Log.getStackTraceString(error));
                        writer.write('\n');
                    }
                    writer.flush();
                }
            }
        } catch (Exception e) {
            Log.e(TAG, "Critical file log write failed", e);
        }
    }

    private static void appendLineToProxySegment(String level, String message, Throwable error) {
        try {
            File logsRoot = resolveLogsRoot();
            if (logsRoot == null) {
                return;
            }
            File poolDir = new File(logsRoot, "pool");
            if (!poolDir.exists() && !poolDir.mkdirs()) {
                Log.e(TAG, "Failed to create proxy pool log directory: " + poolDir.getAbsolutePath());
                return;
            }
            String segmentName = buildProxySegmentName(System.currentTimeMillis());
            File targetFile = new File(poolDir, segmentName + "_proxy.txt");
            synchronized (FILE_LOCK) {
                rotateIfNeeded(targetFile);
                try (FileOutputStream stream = new FileOutputStream(targetFile, true);
                     OutputStreamWriter writer = new OutputStreamWriter(stream, UTF8)) {
                    writer.write(TS_FORMAT.format(new Date()));
                    writer.write(" [");
                    writer.write(level);
                    writer.write("] ");
                    writer.write(message);
                    writer.write('\n');
                    if (error != null) {
                        writer.write(Log.getStackTraceString(error));
                        writer.write('\n');
                    }
                    writer.flush();
                }
            }
        } catch (Exception e) {
            Log.e(TAG, "Proxy segment log write failed", e);
        }
    }

    private static void rotateIfNeeded(File targetFile) {
        if (!targetFile.exists() || targetFile.length() < MAX_FILE_BYTES) {
            return;
        }
        for (int i = MAX_ROTATIONS; i >= 1; i--) {
            File src = i == 1
                    ? targetFile
                    : new File(targetFile.getParentFile(), targetFile.getName() + "." + (i - 1));
            File dst = new File(targetFile.getParentFile(), targetFile.getName() + "." + i);
            if (!src.exists()) {
                continue;
            }
            if (dst.exists() && !dst.delete()) {
                Log.w(TAG, "Failed to delete old rotated file: " + dst.getAbsolutePath());
            }
            if (!src.renameTo(dst)) {
                Log.w(TAG, "Failed to rotate file: " + src.getAbsolutePath());
            }
        }
    }

    private static File resolveLogsRoot() {
        File logs = AppVars.getLogsDir();
        if (logs != null) {
            return logs;
        }
        if (AppVars.getContext() == null) {
            return null;
        }
        File fallback = new File(AppVars.getContext().getFilesDir(), "Logs");
        if (!fallback.exists() && !fallback.mkdirs()) {
            return null;
        }
        return fallback;
    }

    private static String sanitizeChain(String chain) {
        String value = chain.trim().toLowerCase(Locale.ROOT);
        value = value.replaceAll("[^a-z0-9_\\-]", "_");
        if (value.isEmpty()) {
            return "critical";
        }
        return value;
    }

    private static String sanitizeMessage(String message) {
        String value = message.replace('\n', ' ').replace('\r', ' ').trim();
        if (value.length() > 3000) {
            return value.substring(0, 3000) + "...";
        }
        return value;
    }

    private static String buildProxySegmentName(long nowMs) {
        Date date = new Date(nowMs);
        String hourPrefix;
        synchronized (PROXY_SEGMENT_FORMAT) {
            hourPrefix = PROXY_SEGMENT_FORMAT.format(date);
        }
        int minute = Integer.parseInt(new SimpleDateFormat("mm", Locale.US).format(date));
        int bucket = (minute / 10) * 10;
        return hourPrefix + "_" + String.format(Locale.US, "%02d", bucket);
    }
}

File: ANClient/MyHelpers/FileLogger.cs
using System;
using System.IO;
using System.Text;
using System.Threading;
using System.Windows.Forms;

namespace ANClient.MyHelpers
{
    /// <summary>
    /// Файловое логирование — аналог Android FileLogger.java.
    /// 
    /// Пишет отладочные файлы в папку Logs/ рядом с исполняемым файлом.
    /// Структура папок (аналог Android files/Logs/):
    ///   Logs/Critical/{сегмент}_{chain}.log — основные цепочки
    ///   Logs/pool/{сегмент}_proxy.txt — proxy-трафик
    /// 
    /// Сегментация: 10-минутные сегменты (аналог Android SEGMENT_FORMAT).
    /// Ротация: макс. 8 МБ на файл, до 2 бакапов (.1, .2).
    /// 
    /// Потокобезопасность: Single-threaded Executor (аналог Android IO ExecutorService),
    ///   все записи сериализуются через единственный фоновый поток.
    /// 
    /// Зависимости: Application.StartupPath (папка Logs/).
    /// </summary>
    internal static class FileLogger
    {
        private const long MaxFileBytes = 8L * 1024L * 1024L;
        private const int MaxRotations = 2;
        private static readonly object FileLock = new object();

        // Single-threaded queue — аналог Android IO ExecutorService
        private static readonly Thread WriterThread;
        private static readonly AutoResetEvent WakeEvent = new AutoResetEvent(false);
        private static volatile bool _shutdown = false;

        static FileLogger()
        {
            WriterThread = new Thread(WriterLoop) { IsBackground = true, Name = "an-file-logger" };
            WriterThread.Start();
        }

        private static readonly ProducerConsumerQueue Queue = new ProducerConsumerQueue();

        private class LogEntry
        {
            internal string Level;
            internal string Chain;
            internal string Message;
            internal Exception Error;
            internal bool IsProxy;
        }

        private class ProducerConsumerQueue
        {
            private readonly object _lock = new object();
            private readonly System.Collections.Generic.Queue<LogEntry> _items = new System.Collections.Generic.Queue<LogEntry>();

            internal void Enqueue(LogEntry entry)
            {
                lock (_lock)
                {
                    _items.Enqueue(entry);
                }
            }

            internal bool Dequeue(out LogEntry entry)
            {
                lock (_lock)
                {
                    if (_items.Count == 0)
                    {
                        entry = null;
                        return false;
                    }
                    entry = _items.Dequeue();
                    return true;
                }
            }

            internal int Count
            {
                get
                {
                    lock (_lock) { return _items.Count; }
                }
            }
        }

        private static void WriterLoop()
        {
            while (!_shutdown)
            {
                try
                {
                    WakeEvent.WaitOne(5000, false);
                    LogEntry entry;
                    while (Queue.Dequeue(out entry))
                    {
                        try
                        {
                            AppendLineToFile(entry);
                        }
                        catch
                        {
                        }
                    }
                }
                catch
                {
                }
            }
        }

        internal static void Trace(string chain, string message)
        {
            Write("TRACE", chain, message, null);
        }

        internal static void Log(string message)
        {
            Trace("FileLogger", message);
        }

        internal static void Warn(string chain, string message)
        {
            Write("WARN", chain, message, null);
        }

        internal static void Error(string chain, string message, Exception error)
        {
            Write("ERROR", chain, message, error);
        }

        internal static void ProxyPool(string message)
        {
            WriteToProxySegment("TRACE", message, null);
        }

        internal static void ProxyPoolError(string message, Exception error)
        {
            WriteToProxySegment("ERROR", message, error);
        }

        private static void Write(string level, string chain, string message, Exception error)
        {
            if (string.IsNullOrEmpty(chain) || string.IsNullOrEmpty(message))
                return;

            Queue.Enqueue(new LogEntry
            {
                Level = level ?? "TRACE",
                Chain = SanitizeChain(chain),
                Message = SanitizeMessage(message),
                Error = error
            });
            WakeEvent.Set();
        }

        private static void WriteToProxySegment(string level, string message, Exception error)
        {
            if (string.IsNullOrEmpty(message))
                return;

            Queue.Enqueue(new LogEntry
            {
                Level = level ?? "TRACE",
                Chain = "proxy",
                Message = SanitizeMessage(message),
                Error = error,
                IsProxy = true
            });
            WakeEvent.Set();
        }

        private static void AppendLineToFile(LogEntry entry)
        {
            try
            {
                var logsRoot = ResolveLogsRoot();
                if (logsRoot == null)
                    return;

                string targetFile;
                if (entry.IsProxy)
                {
                    var poolDir = Path.Combine(logsRoot, "pool");
                    if (!Directory.Exists(poolDir))
                        Directory.CreateDirectory(poolDir);
                    var segment = BuildSegmentName(DateTime.Now);
                    targetFile = Path.Combine(poolDir, segment + "_proxy.txt");
                }
                else
                {
                    var criticalDir = Path.Combine(logsRoot, "Critical");
                    if (!Directory.Exists(criticalDir))
                        Directory.CreateDirectory(criticalDir);
                    var segment = BuildSegmentName(DateTime.Now);
                    targetFile = Path.Combine(criticalDir, segment + "_" + entry.Chain + ".log");
                }

                RotateIfNeeded(targetFile);

                var line = string.Format("{0} [{1}] {2}{3}",
                    DateTime.Now.ToString("yyyy-MM-dd HH:mm:ss.fff"),
                    entry.Level,
                    entry.Message,
                    Environment.NewLine);

                if (entry.Error != null)
                {
                    line += entry.Error.ToString() + Environment.NewLine;
                }

                lock (FileLock)
                {
                    File.AppendAllText(targetFile, line, Encoding.UTF8);
                }
            }
            catch
            {
            }
        }

        private static string ResolveLogsRoot()
        {
            try
            {
                var path = Path.Combine(Application.StartupPath, "Logs");
                if (!Directory.Exists(path))
                    Directory.CreateDirectory(path);
                return path;
            }
            catch
            {
                return null;
            }
        }

        private static void RotateIfNeeded(string filePath)
        {
            try
            {
                if (!File.Exists(filePath))
                    return;

                var info = new FileInfo(filePath);
                if (info.Length < MaxFileBytes)
                    return;

                for (var i = MaxRotations; i >= 1; i--)
                {
                    var src = i == 1 ? filePath : filePath + "." + (i - 1);
                    var dst = filePath + "." + i;
                    if (!File.Exists(src))
                        continue;
                    if (File.Exists(dst))
                        File.Delete(dst);
                    File.Move(src, dst);
                }
            }
            catch
            {
            }
        }

        private static string SanitizeChain(string chain)
        {
            var value = chain.Trim().ToLowerInvariant();
            foreach (var c in System.IO.Path.GetInvalidFileNameChars())
                value = value.Replace(c, '_');
            if (string.IsNullOrEmpty(value))
                value = "critical";
            return value;
        }

        private static string SanitizeMessage(string message)
        {
            var value = message.Replace('\n', ' ').Replace('\r', ' ').Trim();
            if (value.Length > 3000)
                value = value.Substring(0, 3000) + "...";
            return value;
        }

        private static string BuildSegmentName(DateTime now)
        {
            var hourPrefix = now.ToString("yyyyMMdd_HH");
            var bucket = (now.Minute / 10) * 10;
            return hourPrefix + "_" + bucket.ToString("00");
        }
    }
}
